src/documentation.jl

```diff
@@ -0,0 +1,239 @@
+
+To avoid a `MatchFailure` exception, write the `@match` to handle every possible input.
+One way to do that is to add a final case with the wildcard pattern `_`.
+
+# See Also
+
+See also
+
+- `@match_fail`
+- `@match_return`
+- `@ismatch`
+
+# Patterns:
+
+The following syntactic forms can be used in patterns:
+
+* `_` matches anything
+* `x` (an identifier) matches anything, binds value to the variable `x`
+* `T(x,y,z)` matches structs of type `T` with fields matching patterns `x,y,z`
+* `T(y=1)` matches structs of type `T` whose `y` field equals `1`
+* `[x,y,z]` matches `AbstractArray`s with 3 entries matching `x,y,z`
+* `(x,y,z)` matches `Tuple`s with 3 entries matching `x,y,z`
+* `[x,y...,z]` matches `AbstractArray`s with at least 2 entries, where `x` matches the first entry, `z` matches the last entry and `y` matches the remaining entries.
+* `(x,y...,z)` matches `Tuple`s with at least 2 entries, where `x` matches the first entry, `z` matches the last entry and `y` matches the remaining entries.
+* `::T` matches any subtype (`isa`) of type `T`
+* `x::T` matches any subtype (`isa`) of T that also matches pattern `x`
+* `x || y` matches values which match either pattern `x` or `y` (only variables which exist in both branches will be bound)
+* `x && y` matches values which match both patterns `x` and `y`
+* `x, if condition end` matches only if `condition` is true (`condition` may use any variables that occur earlier in the pattern eg `(x, y, z where x + y > z)`)
+* `x where condition` An alternative form for `x, if condition end`
+* Anything else is treated as a constant and tested for equality
+* Expressions can be interpolated in as constants via standard interpolation syntax `\$(x)`  Interpolations may use previously bound variables.
+
+Patterns can be nested arbitrarily.
+
+Repeated variables only match if they are equal (`isequal`). For example `(x,x)` matches `(1,1)` but not `(1,2)`.
+
+# Examples
+```julia-repl
+julia> value=(1, 2, 3, 4)
+(1, 2, 3, 4)
+
+julia> @match (x, y..., z) = value
+(1, 2, 3, 4)
+
+julia> x
+1
+
+julia> y
+(2, 3)
+
+julia> z
+4
+
+julia> struct Foo
+           x::Int64
+           y::String
+       end
+
+julia> f(x) = @match x begin
+           _::String => :string
+           [a,a,a] => (:all_the_same, a)
+           [a,bs...,c] => (:at_least_2, a, bs, c)
+           Foo(x, "foo") where x > 1 => :foo
+       end
+f (generic function with 1 method)
+
+julia> f("foo")
+:string
+
+julia> f([1,1,1])
+(:all_the_same, 1)
+
+julia> f([1,1])
+(:at_least_2, 1, Int64[], 1)
+
+julia> f([1,2,3,4])
+(:at_least_2, 1, [2, 3], 4)
+
+julia> f([1])
+ERROR: MatchFailure([1])
+...
+
+julia> f(Foo(2, "foo"))
+:foo
+
+julia> f(Foo(0, "foo"))
+ERROR: MatchFailure(Foo(0, "foo"))
+...
+
+julia> f(Foo(2, "not a foo"))
+ERROR: MatchFailure(Foo(2, "not a foo"))
+...
+```
+"""
+macro match end
+
+"""
+    @match_return value
+
+Inside the result part of a @match case, you can return a given value early.
+
+# Examples
```

Comment:
a `documentation.jl` file isn't something I've come across before. Typically folks add the docstrings on the regular declaration site (either in a themed file, or in the main `Package.jl` file

Reply:
Unfortunately, there are several places where different versions of the macros are defined, and the documentation just gets concatenated when you do that.  This ensures we have a single documentation form for each name.